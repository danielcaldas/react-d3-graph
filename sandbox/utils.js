/*eslint require-jsdoc: 0, valid-jsdoc: 0, no-undef: 0, no-empty: 0, no-console: 0*/
import queryString from 'query-string';

/**
 * This two functions generate the react-jsonschema-form
 * schema from some passed graph configuration.
 */
function formMap(k, v) {
    return {
        title: k,
        type: typeof v,
        default: v
    };
}

function generateFormSchema(o, rootSpreadProp, accum = {}) {
    for (let k of Object.keys(o)) {
        const kk = rootSpreadProp ? `${rootSpreadProp}.${k}` : k;

        typeof o[k] === 'object' ? generateFormSchema(o[kk], kk, accum) : (accum[kk] = formMap(kk, o[k]));
    }

    return accum;
}

function loadDataset() {
    const queryParams = queryString.parse(location.search);

    if (queryParams && queryParams.data) {
        const dataset = queryParams.data.toLowerCase();

        try {
            const data = require(`./data/${dataset}/${dataset}.data`);
            const config = require(`./data/${dataset}/${dataset}.config`);

            return { data, config };
        } catch (error) {
            console.warn(
                `dataset with name ${dataset} not found, falling back to default, make sure it is a valid dataset`
            );
        }
    }

    const config = {};
    const data = require('./data/default');

    return {
        config,
        data
    };
}

function setValue(obj, access, value) {
    if (typeof access == 'string') {
        access = access.split('.');
    }

    // check for non existence of root property before advancing
    if (!obj[access[0]]) {
        obj[access[0]] = {};
    }

    access.length > 1 ? setValue(obj[access.shift()], access, value) : (obj[access[0]] = value);
}

export default {
    generateFormSchema,
    loadDataset,
    setValue
};
