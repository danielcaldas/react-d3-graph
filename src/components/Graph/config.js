/**
 * <div style='text-align: right;'><i>This is certainly the only extra piece of documentation that you will ever need</i></div>
 * <br/><br/>
 * Here you can consult a detailed description of each graph configurable property as well as the default values
 * of those properties.
 *
 * **Note about performance**<br/>
 * Some of the properties have a major performance impact when toggled while rendering graphs of medium or large dimensions (hundreds or thousand of elements).
 * These properties are marked with 🚅🚅🚅.<br/>
 * ⭐ **tip** *to achieve smoother interactions you may want to set **staticGraph** to **true** *<br/>
 * <br/>
 * **Note about granularity**<br/>
 * Some of the properties listed in the {@link #node-section|Node section} are marked with 🔍🔍🔍. This means that this properties
 * have a higher level of granularity. These properties can be defined in the graph payload at a node level. (sample payload below)
 * ```javascript
 * const graph = {
 *     nodes: [
 *         {
 *             id: 'id',
 *             color: 'red',         // only this node will be red
 *             size: 300,            // only this node will have size 300
 *             symbolType: 'diamond' // only this node will have 'diamond' shape
 *         }
 *     ],
 *     links: [...]
 * };
 * ```
 *
 * <h2>Graph global configurations</h2>
 * @param {boolean} [automaticRearrangeAfterDropNode=false] - 🚅🚅🚅 when true performing a node drag and drop should automatically
 * rearrange all nodes positions based on new position of dragged node (note: **staticGraph** should be false).
 * @param {number} [height=400] - the height of the (svg) area where the graph will be rendered.
 * @param {boolean} [highlightBehavior=false] - 🚅🚅🚅 when user mouse hovers a node that node and adjacent common
 * connections will be highlighted. All the remaining nodes and links assume opacity value equal to **highlightOpacity**.
 * @param {number} [highlightDegree=1] - **Possible values: 0, 1 or 2**. This value represents the range of the
 * highlight behavior when some node is highlighted. If the value is set to **0** only the selected node will be
 * highlighted. If the value is set to **1** the selected node and his 1st degree connections will be highlighted. If
 * the value is set to **2** the selected node will be highlighted as well as the 1st and 2nd common degree connections.
 * @param {number} [highlightOpacity=1] - this value is used to highlight nodes in the network. The lower
 * the value the more the less highlighted nodes will be visible (related to **highlightBehavior**).
 * @param {number} [maxZoom=8] - max zoom that can be performed against the graph.
 * @param {number} [minZoom=0.1] - min zoom that can be performed against the graph.
 * @param {boolean} [panAndZoom=false] - 🚅🚅🚅 pan and zoom effect when performing zoom in the graph,
 * a similar functionality may be consulted {@link https://bl.ocks.org/mbostock/2a39a768b1d4bc00a09650edef75ad39|here}.
 * @param {boolean} [staticGraph=false] - when setting this value to true the graph will be completely static, thus
 * all forces and drag events upon nodes will produce not effect. Note that, if this value is true the nodes will be
 * rendered with the initial provided **x and y coordinates** (links positions will be automatically set
 * from the given nodes positions by rd3g), no coordinates will be calculated by rd3g or subjacent d3 modules.
 * @param {number} [width=800] - the width of the (svg) area where the graph will be rendered.
 * <br/>
 * @param {Object} node node object is explained in next section.
 * <h2 id="node-section">Node level configurations</h2>
 * @param {string} [node.color='#d3d3d3'] - 🔍🔍🔍 this is the color that will be applied to the node if no **color property**
 * is found inside the node itself (yes **you can pass a property 'color' inside the node and that color will override the
 * this default one**).
 * @param {number} [node.fontSize=10] - {@link https://developer.mozilla.org/en-US/docs/Web/CSS/font-size?v=control|font-size}
 * property for all nodes' labels.
 * @param {string} [node.fontWeight='normal'] - {@link https://developer.mozilla.org/en/docs/Web/CSS/font-weight?v=control|font-weight}
 * property for all nodes' labels.
 * @param {string} [node.labelProperty='id'] - this is the node property that will be used in runtime to
 * fetch the label content. You just need to add some property (e.g. firstName) to the node payload and then set
 * node.labelProperty to be **'firstName'**.
 * @param {string} [node.mouseCursor='pointer'] - {@link https://developer.mozilla.org/en/docs/Web/CSS/cursor?v=control|cursor}
 * property for when some node is mouse hovered.
 * @param {number} [node.opacity=1] - by default all nodes will have this opacity value.
 * @param {boolean} [node.renderLabel=true] - when set to false no labels will appear along side nodes in the
 * graph.
 * @param {number} [node.size=200] - 🔍🔍🔍 defines the size of all nodes.
 * @param {string} [node.strokeColor='none'] - color for the stroke of each node.
 * @param {number} [node.strokeWidth=1.5] - the width of the all node strokes.
 * @param {string} [node.symbolType='circle'] - 🔍🔍🔍 the <span id='node-symbol-type'>shape</span> of the node.
 * Use the following values under a property **type** inside each node (nodes may have different types, same as colors):
 *   - 'circle'
 *   - 'cross'
 *   - 'diamond'
 *   - 'square'
 *   - 'star'
 *   - 'triangle'
 *   - 'wye'
 *
 * **[note]** react-d3-graph will map this values to d3 symbols ({@link https://github.com/d3/d3-shape#symbols})
 * @param {string} [node.highlightColor='SAME'] - color for all highlighted nodes (use string 'SAME' if you
 * want the node to keep its color in highlighted state).
 * @param {number} [node.highlightFontSize=10] - node.fontSize in highlighted state.
 * @param {string} [node.highlightFontWeight='normal'] - node.fontWeight in highlighted state.
 * @param {string} [node.highlightStrokeColor='SAME'] - node.strokeColor in highlighted state.
 * @param {number} [node.highlightStrokeWidth=1.5] - node.strokeWidth in highlighted state.
 * <br/>
 * @param {Object} link link object is explained in the next section.
 * <h2>Link level configurations</h2>
 * @param {string} [link.color='#d3d3d3'] - the color for links.
 * @param {number} [link.opacity=1] - the default opacity value for links.
 * @param {boolean} [link.semanticStrokeWidth=false] - when set to true all links will have
 * *"semantic width"*, this means that the width of the connections will be proportional to the value of each link.
 * This is how link strokeWidth will be calculated:
 * ```javascript
 * strokeWidth += (linkValue * strokeWidth) / 10;
 * ```
 * @param {number} [link.strokeWidth=1.5] - strokeWidth for all links.
 * @param {string} [link.highlightColor='#d3d3d3'] - links color in highlight state.
 *
 * @example
 * // A simple config that uses some properties
 * const myConfig = {
 *     highlightBehavior: true,
 *     node: {
 *         color: 'lightgreen',
 *         size: 120,
 *         highlightStrokeColor: 'blue'
 *     },
 *     link: {
 *         highlightColor: 'lightblue'
 *     }
 * };
 *
 * // Sorry for the long config description, here's a potato 🥔.
 */
export default {
    automaticRearrangeAfterDropNode: false,
    height: 400,
    highlightBehavior: true,
    highlightDegree: 1,
    highlightOpacity: 0.1,
    maxZoom: 8,
    minZoom: 0.5,
    panAndZoom: false,
    staticGraph: false,
    width: 800,
    node: {
        color: '#d3d3d3',
        fontSize: 8,
        fontWeight: 'normal',
        labelProperty: 'id',
        mouseCursor: 'pointer',
        opacity: 1,
        renderLabel: true,
        size: 80,
        strokeColor: 'none',
        strokeWidth: 1.5,
        symbolType: 'circle',
        highlightColor: 'SAME',
        highlightFontSize: 8,
        highlightFontWeight: 'normal',
        highlightStrokeColor: 'SAME',
        highlightStrokeWidth: 1.5
    },
    link: {
        color: '#d3d3d3',
        opacity: 1,
        semanticStrokeWidth: false,
        strokeWidth: 1.5,
        highlightColor: '#d3d3d3'
    }
};
