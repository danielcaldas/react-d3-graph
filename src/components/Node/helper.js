/**
 * @module Node/helper
 * @description
 * Some methods that help no the process of rendering a node.
 */
import * as d3 from 'd3';

import CONST from './const';

/**
 * Converts a string that specifies a symbol into a concrete instance
 * of d3 symbol.<br/>
 * {@link https://github.com/d3/d3-shape/blob/master/README.md#symbol}
 * @param  {string} typeName - the string that specifies the symbol type (should be one of {@link #node-symbol-type|node.symbolType}).
 * @return {Object} concrete instance of d3 symbol (defaults to circle).
 * @memberof Node/helper
 */
function _convertTypeToD3Symbol(typeName) {
    switch (typeName) {
        case CONST.SYMBOLS.CIRCLE:
            return d3.symbolCircle;
        case CONST.SYMBOLS.CROSS:
            return d3.symbolCross;
        case CONST.SYMBOLS.DIAMOND:
            return d3.symbolDiamond;
        case CONST.SYMBOLS.SQUARE:
            return d3.symbolSquare;
        case CONST.SYMBOLS.STAR:
            return d3.symbolStar;
        case CONST.SYMBOLS.TRIANGLE:
            return d3.symbolTriangle;
        case CONST.SYMBOLS.WYE:
            return d3.symbolWye;
        default:
            return d3.symbolCircle;
    }
}

/**
 * Build a d3 svg symbol based on passed symbol and symbol type.
 * @param  {number} [size=80] - the size of the symbol.
 * @param  {string} [symbolTypeDesc='circle'] - the string containing the type of symbol that we want to build
 * (should be one of {@link #node-symbol-type|node.symbolType}).
 * @return {Object} concrete instance of d3 symbol.
 * @memberof Node/helper
 */
function buildSvgSymbol(size=80, symbolTypeDesc=CONST.SYMBOLS.CIRCLE) {
    return d3.symbol()
            .size(() => size)
            .type(() => _convertTypeToD3Symbol(symbolTypeDesc))(); // @todo: Strange behavior d3.Symbol ret function
}

export default {
    buildSvgSymbol
};
