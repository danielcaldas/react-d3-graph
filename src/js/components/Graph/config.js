export default {
    automaticRearrangeAfterDropNode: false, // Very expensive for dense graphs, this will start a new force ...
    height: 450,
    highlightBehavior: false,
    highlightOpacity: 1, // For all elements: nodes, text and links
    maxZoom: 8,
    minZoom: 0.5,
    staticGraph: false,
    width: 800,
    node: {
        color: '#d3d3d3',
        fontSize: 10, // in px
        fontWeight: 'normal',
        highlightColor: 'SAME',
        highlightFontSize: 10,
        highlightFontWeight: 'normal',
        highlightStrokeColor: 'none',
        highlightStrokeWidth: 1.5,
        labelProperty: 'id',
        mouseCursor: 'pointer',
        opacity: 1,
        renderLabel: true,
        size: 200,
        strokeColor: 'none',
        strokeWidth: 1.5,
        symbolType: 'circle'
    },
    link: {
        color: '#d3d3d3',
        highlightColor: '#d3d3d3',
        opacity: 1,
        strokeWidth: 1.5
    }
};
