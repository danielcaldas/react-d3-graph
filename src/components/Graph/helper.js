import React from 'react';

import * as d3 from 'd3';

import CONST from './const';

import Link from '../Link/';
import Node from '../Node/';

function buildGraph(nodes, nodeCallbacks, links, linkCallbacks, config, someNodeHighlighted) {
    let linksComponents = [];
    let nodesComponents = [];

    for (let node of Object.values(nodes)) {
        const props = buildNodeProps(node, config, nodeCallbacks, someNodeHighlighted);

        nodesComponents.push(<Node key={node.id} {...props} />);

        linksComponents = linksComponents.concat(buildNodeLinks(node, nodes, links, config, linkCallbacks, someNodeHighlighted));
    }

    return {
        nodes: nodesComponents,
        links: linksComponents
    };
}

function buildNodeProps(node, config, nodeCallbacks, someNodeHighlighted) {
    const opacity = someNodeHighlighted ? (node.highlighted ? config.node.opacity : config.highlightOpacity) : config.node.opacity;

    let fill = node.color || config.node.color;

    if (node.highlighted && config.node.highlightColor !== CONST.KEYWORDS.SAME) {
        fill = config.node.highlightColor;
    }

    return {
        className: CONST.NODE_CLASS_NAME,
        cursor: config.node.mouseCursor,
        cx: node && node.x || '0',
        cy: node && node.y || '0',
        fill,
        fontSize: node.highlighted ? config.node.highlightFontSize : config.node.fontSize,
        fontWeight: node.highlighted ? config.node.highlightFontWeight : config.node.fontWeight,
        id: node.id,
        label: node[config.node.labelProperty] || node.id,
        onClickNode: nodeCallbacks.onClickNode,
        onMouseOverNode: nodeCallbacks.onMouseOverNode,
        onMouseOut: nodeCallbacks.onMouseOut,
        opacity,
        renderLabel: config.node.renderLabel,
        size: node.size || config.node.size,
        stroke: node.highlighted ? config.node.highlightStrokeColor : config.node.strokeColor,
        strokeWidth: node.highlighted ? config.node.highlightStrokeWidth : config.node.strokeWidth,
        type: node.type || config.node.symbolType
    };
}

function buildNodeLinks(node, nodes, links, config, linkCallbacks, someNodeHighlighted) {
    let linksComponents = [];

    if (links[node.id]) {
        const x1 = node && node.x || '0';
        const y1 = node && node.y || '0';

        const adjacents = Object.keys(links[node.id]);
        const n = adjacents.length;

        for (let j=0; j < n; j++) {
            const target = adjacents[j];

            if (nodes[target]) {
                const key = `${node.id}${CONST.COORDS_SEPARATOR}${target}`;
                const props = buildLinkProps(node.id, target, x1, y1, nodes, links, config, linkCallbacks, someNodeHighlighted);

                linksComponents.push(<Link key={key} {...props} />);
            }
        }
    }

    return linksComponents;
}

function buildLinkProps(source, target, x1, y1, nodes, links, config, linkCallbacks, someNodeHighlighted) {
    const opacity = someNodeHighlighted ? (nodes[source].highlighted && nodes[target].highlighted) ? config.link.opacity : config.highlightOpacity : config.link.opacity;

    const stroke = (nodes[source].highlighted && nodes[target].highlighted) ?
                (config.link.highlightColor === CONST.KEYWORDS.SAME ? config.link.color : config.link.highlightColor)
                : config.link.color;

    const linkValue = links[source][target] || links[target][source];
    let strokeWidth = config.link.strokeWidth;

    if (config.link.semanticStrokeWidth) {
        strokeWidth += (linkValue * strokeWidth) / 10;
    }

    return {
        source,
        target,
        x1,
        y1,
        x2: nodes[target] && nodes[target].x || '0',
        y2: nodes[target] && nodes[target].y || '0',
        strokeWidth,
        stroke,
        className: CONST.LINK_CLASS_NAME,
        opacity,
        onClickLink: linkCallbacks.onClickLink
    };
}

function createForceSimulation(width, height) {
    const forceX = d3.forceX(width / 2).strength(CONST.FORCE_X);
    const forceY = d3.forceY(height / 2).strength(CONST.FORCE_Y);

    const simulation = d3.forceSimulation()
            .force('charge', d3.forceManyBody().strength(CONST.FORCE_IDEAL_STRENGTH))
            .force('x', forceX)
            .force('y', forceY);

    return simulation;
}

function initializeNodes(graphNodes) {
    let nodes = {};
    let indexMapping = {};
    let index = 0;

    graphNodes.forEach(n => {
        n['highlighted'] = false;
        if (!n.hasOwnProperty('x')) n['x'] = 0;
        if (!n.hasOwnProperty('y')) n['y'] = 0;

        nodes[n.id.toString()] = n;
        indexMapping[index] = n.id;

        index++;
    });

    return {
        nodes,
        indexMapping
    };
}

function initializeLinks(graphLinks) {
    let links = {};

    graphLinks.forEach(l => {
        if (!links[l.source]) {
            links[l.source] = {};
        }

        if (!links[l.target]) {
            links[l.target] = {};
        }

        // @TODO: If the graph is directed this should not happen
        links[l.source][l.target] = links[l.target][l.source] = l.value || 1;
    });

    return links;
}

export default {
    buildGraph,
    createForceSimulation,
    initializeLinks,
    initializeNodes
};
