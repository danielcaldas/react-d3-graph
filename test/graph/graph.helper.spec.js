import * as graphHelper from "../../src/components/graph/graph.helper";

import config from "../../src/components/graph/graph.config";

import utils from "../../src/utils";
import * as linkHelper from "../../src/components/link/link.helper";

jest.mock("d3-force");
import {
    forceX as d3ForceX,
    forceY as d3ForceY,
    forceSimulation as d3ForceSimulation,
    forceManyBody as d3ForceManyBody,
} from "d3-force";

describe("Graph Helper", () => {
    beforeAll(() => {
        utils.isDeepEqual = jest.fn();
        utils.isEmptyObject = jest.fn();
        utils.merge = jest.fn();
        utils.throwErr = jest.fn();
        jest.spyOn(linkHelper, "buildLinkPathDefinition");
    });

    describe("#buildLinkProps", () => {
        let that = {};

        beforeAll(() => {
            that = {
                config: { link: config.link },
                link: { source: "source", target: "target" },
            };
        });

        describe("when building props for a link", () => {
            test("should call buildLinkPathDefinition with expected parameters", () => {
                graphHelper.buildLinkProps(that.link, {}, {}, that.config, [], undefined, undefined, 1);

                expect(linkHelper.buildLinkPathDefinition).toHaveBeenCalledWith(
                    { source: { x: 0, y: 0 }, target: { x: 0, y: 0 } },
                    "STRAIGHT"
                );
            });

            describe("and no custom color is set", () => {
                test("should return default color defined in link config", () => {
                    const props = graphHelper.buildLinkProps(
                        that.link,
                        {},
                        {},
                        that.config,
                        [],
                        undefined,
                        undefined,
                        1
                    );

                    expect(props.stroke).toEqual(that.config.link.color);
                });
            });

            describe("and custom color is set to green", () => {
                test("should return green color in the props", () => {
                    const props = graphHelper.buildLinkProps(
                        { ...that.link, color: "green" },
                        {},
                        {},
                        that.config,
                        [],
                        undefined,
                        undefined,
                        1
                    );

                    expect(props.stroke).toEqual("green");
                });
            });
        });
    });

    describe("#buildNodeProps", () => {
        let that = {};

        beforeEach(() => {
            const nodeConfig = Object.assign({}, config.node, { svg: "file.svg" });
            const linkConfig = Object.assign({}, config.link);

            that = {
                config: { node: nodeConfig, link: linkConfig },
                node: {
                    id: "id",
                    x: 1,
                    y: 2,
                    color: "green",
                    highlighted: false,
                    symbolType: undefined,
                },
            };
        });
        describe("when node to build is the highlightedNode", () => {
            test("should return node props with proper highlight values", () => {
                that.node.highlighted = true;
                Object.assign(that.config.node, {
                    highlightColor: "red",
                    highlightFontSize: 12,
                    highlightFontWeight: "bold",
                    highlightStrokeColor: "yellow",
                    highlightStrokeWidth: 2,
                });
                const props = graphHelper.buildNodeProps(that.node, that.config, undefined, "id", undefined, 1);

                expect(props).toEqual({
                    ...that.node,
                    className: "node",
                    cursor: "pointer",
                    cx: 1,
                    cy: 2,
                    dx: 15.5,
                    fill: "red",
                    fontSize: 12,
                    fontWeight: "bold",
                    fontColor: "black",
                    id: "id",
                    label: "id",
                    onClickNode: undefined,
                    onRightClickNode: undefined,
                    onMouseOut: undefined,
                    onMouseOverNode: undefined,
                    opacity: 1,
                    renderLabel: true,
                    size: 200,
                    stroke: "yellow",
                    strokeWidth: 2,
                    svg: "file.svg",
                    type: "circle",
                    viewGenerator: null,
                    overrideGlobalViewGenerator: undefined,
                });
            });
        });
        describe("when node to build is the highlightedLink target (or source)", () => {
            describe("and highlight degree is 0", () => {
                test("should properly build node props ()", () => {
                    that.config.highlightDegree = 0;

                    const props = graphHelper.buildNodeProps(
                        that.node,
                        that.config,
                        undefined,
                        {
                            source: "some other id",
                            target: "id",
                        },
                        undefined,
                        1
                    );

                    expect(props).toEqual({
                        ...that.node,
                        className: "node",
                        cursor: "pointer",
                        cx: 1,
                        cy: 2,
                        dx: 11.5,
                        fill: "green",
                        fontSize: 8,
                        fontWeight: "normal",
                        fontColor: "black",
                        id: "id",
                        label: "id",
                        onClickNode: undefined,
                        onRightClickNode: undefined,
                        onMouseOut: undefined,
                        onMouseOverNode: undefined,
                        opacity: undefined,
                        renderLabel: true,
                        size: 200,
                        stroke: "none",
                        strokeWidth: 1.5,
                        svg: "file.svg",
                        type: "circle",
                        viewGenerator: null,
                        overrideGlobalViewGenerator: undefined,
                    });
                });
            });
            describe("and highlight degree is bigger then 0", () => {
                test("should properly build node props", () => {
                    that.config.highlightDegree = 2;

                    const props = graphHelper.buildNodeProps(
                        that.node,
                        that.config,
                        undefined,
                        {
                            source: "some other id",
                            target: "id",
                        },
                        undefined,
                        1
                    );

                    expect(props).toEqual({
                        ...that.node,
                        className: "node",
                        cursor: "pointer",
                        cx: 1,
                        cy: 2,
                        dx: 11.5,
                        fill: "green",
                        fontSize: 8,
                        fontWeight: "normal",
                        fontColor: "black",
                        id: "id",
                        label: "id",
                        onClickNode: undefined,
                        onRightClickNode: undefined,
                        onMouseOut: undefined,
                        onMouseOverNode: undefined,
                        opacity: undefined,
                        renderLabel: true,
                        size: 200,
                        stroke: "none",
                        strokeWidth: 1.5,
                        svg: "file.svg",
                        type: "circle",
                        viewGenerator: null,
                        overrideGlobalViewGenerator: undefined,
                    });
                });
            });
        });
        describe("and no custom strokeColor is set", () => {
            test("should return the default strokeColor in the props", () => {
                const props = graphHelper.buildNodeProps(that.node, that.config, undefined, undefined, undefined, 1);

                expect(props.stroke).toEqual("none");
            });
        });
        describe("and custom strokeColor is set to yellow", () => {
            test("should return yellow strokeColor in the props", () => {
                const props = graphHelper.buildNodeProps(
                    { ...that.node, strokeColor: "yellow" },
                    that.config,
                    undefined,
                    undefined,
                    undefined,
                    1
                );

                expect(props.stroke).toEqual("yellow");
            });
        });
    });

    describe("#initializeGraphState", () => {
        describe("when valid graph data is provided", () => {
            beforeEach(() => {
                const fr = 10;
                const forceStub = jest.fn();

                d3ForceX.mockImplementation(() => ({ strength: () => fr }));
                d3ForceY.mockImplementation(() => ({ strength: () => fr }));
                d3ForceManyBody.mockImplementation(() => ({ strength: () => fr }));
                forceStub.mockImplementation(() => ({ force: forceStub }));
                d3ForceSimulation.mockImplementation(() => ({ force: forceStub }));
                utils.merge.mockImplementation(() => ({ config: "config" }));
            });

            describe("and received state was already initialized", () => {
                test("should create graph structure absorbing stored nodes and links behavior", () => {
                    const data = {
                        nodes: [{ id: "A" }, { id: "B" }, { id: "C" }],
                        links: [{ source: "A", target: "B" }, { source: "C", target: "A" }],
                    };
                    const state = {
                        nodes: {
                            A: { x: 20, y: 40 },
                            B: { x: 40, y: 60 },
                        },
                        links: [],
                        nodeIndexMapping: "nodeIndexMapping",
                    };

                    const newState = graphHelper.initializeGraphState({ data, id: "id", config: {} }, state);

                    expect(newState.d3Nodes).toEqual([
                        {
                            highlighted: false,
                            id: "A",
                            x: 20,
                            y: 40,
                        },
                        {
                            highlighted: false,
                            id: "B",
                            x: 40,
                            y: 60,
                        },
                        {
                            highlighted: false,
                            id: "C",
                            x: 0,
                            y: 0,
                        },
                    ]);
                    expect(newState.d3Links).toEqual([
                        {
                            index: 0,
                            source: {
                                highlighted: false,
                                id: "A",
                            },
                            target: {
                                highlighted: false,
                                id: "B",
                            },
                        },
                        {
                            index: 1,
                            source: {
                                highlighted: false,
                                id: "C",
                            },
                            target: {
                                highlighted: false,
                                id: "A",
                            },
                        },
                    ]);
                });
            });

            describe("and received state is empty", () => {
                test("should create new graph structure with nodes and links", () => {
                    const data = {
                        nodes: [{ id: "A" }, { id: "B" }, { id: "C" }],
                        links: [{ source: "A", target: "B" }, { source: "C", target: "A" }],
                    };
                    const state = {};

                    const newState = graphHelper.initializeGraphState({ data, id: "id", config: {} }, state);

                    expect(newState.d3Nodes).toEqual([
                        {
                            highlighted: false,
                            id: "A",
                            x: 0,
                            y: 0,
                        },
                        {
                            highlighted: false,
                            id: "B",
                            x: 0,
                            y: 0,
                        },
                        {
                            highlighted: false,
                            id: "C",
                            x: 0,
                            y: 0,
                        },
                    ]);
                    expect(newState.d3Links).toEqual([
                        {
                            source: "A",
                            target: "B",
                        },
                        {
                            source: "C",
                            target: "A",
                        },
                    ]);
                });
            });

            test("should return proper state object for given inputs", () => {
                const forceStub = jest.fn();

                forceStub.mockImplementation(() => {
                    return {
                        force: forceStub,
                    };
                });

                d3ForceSimulation.mockImplementation(() => {
                    return {
                        force: forceStub,
                    };
                });

                const data = {
                    nodes: [{ id: "A" }, { id: "B" }, { id: "C" }],
                    links: [{ source: "A", target: "B" }, { source: "C", target: "A" }],
                };
                const state = {
                    nodes: {
                        A: { x: 20, y: 40 },
                        B: { x: 40, y: 60 },
                    },
                    links: "links",
                    nodeIndexMapping: "nodeIndexMapping",
                };

                const newState = graphHelper.initializeGraphState({ data, id: "id", config: undefined }, state);

                expect(newState).toEqual({
                    config: {
                        config: "config",
                    },
                    configUpdated: false,
                    d3Links: [
                        {
                            index: 0,
                            source: {
                                highlighted: false,
                                id: "A",
                            },
                            target: {
                                highlighted: false,
                                id: "B",
                            },
                        },
                        {
                            index: 1,
                            source: {
                                highlighted: false,
                                id: "C",
                            },
                            target: {
                                highlighted: false,
                                id: "A",
                            },
                        },
                    ],
                    d3Nodes: [
                        {
                            highlighted: false,
                            id: "A",
                            x: 20,
                            y: 40,
                        },
                        {
                            highlighted: false,
                            id: "B",
                            x: 40,
                            y: 60,
                        },
                        {
                            highlighted: false,
                            id: "C",
                            x: 0,
                            y: 0,
                        },
                    ],
                    highlightedNode: "",
                    id: "id",
                    links: {
                        A: {
                            B: 1,
                            C: 1,
                        },
                        B: {
                            A: 1,
                        },
                        C: {
                            A: 1,
                        },
                    },
                    newGraphElements: false,
                    nodes: {
                        A: {
                            highlighted: false,
                            id: "A",
                            x: 20,
                            y: 40,
                        },
                        B: {
                            highlighted: false,
                            id: "B",
                            x: 40,
                            y: 60,
                        },
                        C: {
                            highlighted: false,
                            id: "C",
                            x: 0,
                            y: 0,
                        },
                    },
                    simulation: {
                        force: forceStub,
                    },
                    transform: 1,
                });
            });
        });

        describe("when invalid graph data is provided", () => {
            describe("when no nodes are provided", () => {
                test("should throw INSUFFICIENT_DATA error", () => {
                    const data = { nodes: [], links: [] };

                    graphHelper.initializeGraphState(
                        {
                            data,
                            id: "id",
                            config: "config",
                        },
                        "state"
                    );

                    expect(utils.throwErr).toHaveBeenCalledWith(
                        "Graph",
                        "you have not provided enough data" +
                            " for react-d3-graph to render something. You need to provide at least one node"
                    );
                });
            });

            describe("when invalid link structure is found", () => {
                afterEach(() => {
                    utils.throwErr.mockReset();
                });

                describe("when link source references nonexistent node", () => {
                    test("should throw INVALID_LINKS error", () => {
                        const data = { nodes: [{ id: "A" }], links: [{ source: "B", target: "A" }] };

                        graphHelper.initializeGraphState(
                            {
                                data,
                                id: "id",
                                config: "config",
                            },
                            "state"
                        );

                        expect(utils.throwErr).toHaveBeenCalledWith(
                            "Graph",
                            "you provided a invalid links data" +
                                // eslint-disable-next-line
                                ' structure. Links source and target attributes must point to an existent node - "B" is not a valid source node id'
                        );
                    });
                });

                describe("when link target references nonexistent node", () => {
                    test("should throw INVALID_LINKS error", () => {
                        const data = { nodes: [{ id: "A" }], links: [{ source: "A", target: "B" }] };

                        graphHelper.initializeGraphState(
                            {
                                data,
                                id: "id",
                                config: "config",
                            },
                            "state"
                        );

                        expect(utils.throwErr).toHaveBeenCalledWith(
                            "Graph",
                            "you provided a invalid links data" +
                                // eslint-disable-next-line
                                ' structure. Links source and target attributes must point to an existent node - "B" is not a valid target node id'
                        );
                    });
                });
            });
        });
    });
});
