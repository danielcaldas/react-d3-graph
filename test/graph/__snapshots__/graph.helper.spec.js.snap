// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Helper #initializeGraphState when valid graph data is provided and received state is empty should create new graph structure with nodes and links 1`] = `
Array [
  Object {
    "highlighted": false,
    "id": "A",
    "x": 0,
    "y": 0,
  },
  Object {
    "highlighted": false,
    "id": "B",
    "x": 0,
    "y": 0,
  },
  Object {
    "highlighted": false,
    "id": "C",
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Graph Helper #initializeGraphState when valid graph data is provided and received state is empty should create new graph structure with nodes and links 2`] = `
Array [
  Object {
    "source": "A",
    "target": "B",
  },
  Object {
    "source": "C",
    "target": "A",
  },
]
`;

exports[`Graph Helper #initializeGraphState when valid graph data is provided and received state was already initialized and the new state has nodes removed should create graph structure preserving subset of original structure 1`] = `
Array [
  Object {
    "highlighted": false,
    "id": "B",
    "x": 40,
    "y": 60,
  },
  Object {
    "highlighted": false,
    "id": "C",
    "x": 60,
    "y": 80,
  },
]
`;

exports[`Graph Helper #initializeGraphState when valid graph data is provided and received state was already initialized and the new state has nodes removed should create graph structure preserving subset of original structure 2`] = `
Array [
  Object {
    "index": 2,
    "isHidden": false,
    "source": Object {
      "highlighted": false,
      "id": "B",
    },
    "target": Object {
      "highlighted": false,
      "id": "C",
    },
  },
]
`;

exports[`Graph Helper #initializeGraphState when valid graph data is provided and received state was already initialized should create graph structure absorbing stored nodes and links behavior 1`] = `
Array [
  Object {
    "highlighted": false,
    "id": "A",
    "x": 20,
    "y": 40,
  },
  Object {
    "highlighted": false,
    "id": "B",
    "x": 40,
    "y": 60,
  },
  Object {
    "highlighted": false,
    "id": "C",
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Graph Helper #initializeGraphState when valid graph data is provided and received state was already initialized should create graph structure absorbing stored nodes and links behavior 2`] = `
Array [
  Object {
    "index": 0,
    "source": Object {
      "highlighted": false,
      "id": "A",
    },
    "target": Object {
      "highlighted": false,
      "id": "B",
    },
  },
  Object {
    "index": 1,
    "source": Object {
      "highlighted": false,
      "id": "C",
    },
    "target": Object {
      "highlighted": false,
      "id": "A",
    },
  },
]
`;
